File: training_1.0/8_trees/task_a.cpp
#include <iostream>
#include <fstream>
using namespace std;

struct Tree {
    int key;
    Tree* left;
    Tree* right;
};

void add(Tree*& root, int key) {
    if (!root) {
        root = new Tree;
        root->key = key;
        root->left = nullptr;
        root->right = nullptr;
    } else {
        if (root->key == key) {
            return;
        } else if (root->key > key) {
            add(root->left, key);
        } else {
            add(root->right, key);
        }
    }
}

int max_depth(Tree*& root) {
    if (!root) return 0;

    int left = max_depth(root->left);
    int right = max_depth(root->right);

    if (left <= right) return right+1;
    else return left+1;
}

int main() {
    ifstream fin("input.txt");
    ofstream fout("output.txt");

    Tree* root = nullptr;

    int n;
    while (fin >> n) {
        if (n != 0) add(root, n);
    }

    fout << max_depth(root);
}

File: training_1.0/7_event sorting/task_a.py
def count_unobserved_tables(n, m, tin, tout):
    tables = []
    for i in range(m):
        tables.append((tin[i], -1))
        tables.append((tout[i], 1))
    tables.sort()
    observed = 0
    unobserved = n
    left = None
    for table in tables:
        if table[1] == -1:
            if observed == 0:
                left = table[0]
            observed += 1
        else:
            observed -= 1
            if observed == 0:
                unobserved -= (table[0] - left + 1)
    return unobserved


fin = open("input.txt", 'r')
fout = open("output.txt", 'w')
n, m = map(int, fin.readline().strip().split())
tin, tout = [0] * m, [0] * m

for i in range(m):
    tin[i], tout[i] = map(int, fin.readline().strip().split())

print(count_unobserved_tables(n, m, tin, tout), file=fout)
